### ISSUE TITLE
BasicsTest#test_readonly_attributes fails depending on test files order

### ISSUE THREAD

**Issue by yahonda** [rails/rails#26368]
### Steps to reproduce
I'm running ActiveRecord unit tests in my local environment. 
```
$ cd activerecord
$ bundle exec rake test
```

Here is the minimum steps to reproduce.
```ruby
for i in sqlite3 mysql2 postgresql
do
ARCONN=$i bundle exec ruby -w -I"lib:test" -I"/home/yahonda/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rake-11.1.2/lib" "/home/yahonda/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rake-11.1.2/lib/rake/rake_test_loader.rb" "test/cases/relation/where_test.rb" "test/cases/base_test.rb"
done
```
It only reproduces test files listed this order. "test/cases/relation/where_test.rb" first then "test/cases/base_test.rb". If these file order flipped like "test/cases/base_test.rb" "test/cases/relation/where_test.rb", it will not reproduce.

### Expected behavior
All unit test finish without failures.

### Actual behavior
`BasicsTest#test_readonly_attributes` fails for all bundled adapters, sqlite3 mysql2 and postgresql.

```ruby
$ for i in sqlite3 mysql2 postgresql
> do
> ARCONN=$i bundle exec ruby -w -I"lib:test" -I"/home/yahonda/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rake-11.1.2/lib" "/home/yahonda/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rake-11.1.2/lib/rake/rake_test_loader.rb" \
> "test/cases/relation/where_test.rb" "test/cases/base_test.rb"
> done
Using sqlite3
Run options: --seed 30872

# Running:

.................................................S.....................F...................................................................................................................

Finished in 1.414296s, 132.2213 runs/s, 289.8970 assertions/s.

  1) Failure:
BasicsTest#test_readonly_attributes [/home/yahonda/git/rails/activerecord/test/cases/base_test.rb:642]:
--- expected
+++ actual
@@ -1 +1 @@
-#<Set: {"title", "comments_count"}>
+#<Set: {"title"}>


187 runs, 410 assertions, 1 failures, 0 errors, 1 skips

You have skipped tests. Run with --verbose for details.
Using mysql2
Run options: --seed 53967

# Running:

........................................................................................................................S..........................................................F........

Finished in 2.064569s, 91.0601 runs/s, 199.0730 assertions/s.

  1) Failure:
BasicsTest#test_readonly_attributes [/home/yahonda/git/rails/activerecord/test/cases/base_test.rb:642]:
--- expected
+++ actual
@@ -1 +1 @@
-#<Set: {"title", "comments_count"}>
+#<Set: {"title"}>


188 runs, 411 assertions, 1 failures, 0 errors, 1 skips

You have skipped tests. Run with --verbose for details.
Using postgresql
Run options: --seed 46193

# Running:

......................................................................................................F...................................................................................

Finished in 2.510094s, 74.1008 runs/s, 164.9340 assertions/s.

  1) Failure:
BasicsTest#test_readonly_attributes [/home/yahonda/git/rails/activerecord/test/cases/base_test.rb:642]:
--- expected
+++ actual
@@ -1 +1 @@
-#<Set: {"title", "comments_count"}>
+#<Set: {"title"}>


186 runs, 414 assertions, 1 failures, 0 errors, 0 skips
$
```

### System configuration
**Rails version**:
master branch 
**Ruby version**:
ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-linux]

**Comment by maclover7** [comment]
Investigating. Looks like the `comment_count` attribute is a counter cache column, was introduced via https://github.com/rails/rails/commit/c48f744400844e4c73eb91c83d86c3e915a9d78b.

**Comment by prathamesh-sonpatki** [comment]
Noticed that tests pass if we move https://github.com/rails/rails/blob/unlock-minitest/activerecord/test/cases/relation/where_test.rb#L10 before https://github.com/rails/rails/blob/unlock-minitest/activerecord/test/cases/relation/where_test.rb#L7

**Comment by yahonda** [comment]
Thanks guys. moving up `require "models/post"`  before `require "models/comment"` works for my minimum testcase executing these two files`"test/cases/relation/where_test.rb" "test/cases/base_test.rb"`.

But `bundle exec rake test` shows the same failure for all bundled adapters. Changing require order might have caused another order dependency.

**Comment by yahonda** [comment]
pull request #26370 has been opened.

**Comment by pixeltrix** [comment]
Fixed by 1f0112dfc38b1044639fed3ca0d6f1c203f97394

**Comment by pixeltrix** [comment]
Is this causing intermittent test failures on 4-2-stable and 5-0-stable branches on Travis?
